modules/compatibility/src/test/java/org/apache/ignite/compatibility/testframework/junits/IgniteCompatibilityAbstractTest.java

```diff
@@ -0,0 +1,331 @@
+                    if (arg.startsWith("-Xmx") || arg.startsWith("-Xms"))
+                        filteredJvmArgs.add(arg);
+                }
+
+                String classPath = System.getProperty("java.class.path");
+
+                String[] paths = classPath.split(File.pathSeparator);
+
+                StringBuilder pathBuilder = new StringBuilder();
+
+                String corePathTemplate = "ignite.modules.core.target.classes".replace(".", File.separator);
+                String coreTestsPathTemplate = "ignite.modules.core.target.test-classes".replace(".", File.separator);
+
+                for (String path : paths) {
+                    if (!path.contains(corePathTemplate) && !path.contains(coreTestsPathTemplate))
+                        pathBuilder.append(path).append(File.pathSeparator);
+                }
+
+                String pathToArtifact = MavenUtils.getPathToIgniteCoreArtifact(ver);
+                pathBuilder.append(pathToArtifact).append(File.pathSeparator);
+
+                String pathToTestsArtifact = MavenUtils.getPathToIgniteCoreArtifact(ver, "tests");
+                pathBuilder.append(pathToTestsArtifact).append(File.pathSeparator);
+
+                filteredJvmArgs.add("-cp");
+                filteredJvmArgs.add(pathBuilder.toString());
+
+                return filteredJvmArgs;
+            }
+        };
+
+        if (locJvmInstance == null) {
+            CountDownLatch nodeJoinedLatch = rmJvmInstance == null ? new CountDownLatch(1) : new CountDownLatch(2);
+
+            UUID nodeId = ignite.getId();
+
+            ListenedGridTestLog4jLogger log = (ListenedGridTestLog4jLogger)ignite.log();
+
+            log.addListener(nodeId, new LoggedJoinNodeClosure(nodeJoinedLatch, nodeId));
+
+            assert nodeJoinedLatch.await(NODE_JOIN_TIMEOUT, TimeUnit.SECONDS) : "Node has not joined [id=" + nodeId + "]";
+
+            log.removeListener(nodeId);
+        }
+
+        if (rmJvmInstance == null)
+            rmJvmInstance = ignite;
+
+        return ignite;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected Ignite startGrid(String igniteInstanceName, IgniteConfiguration cfg,
+        GridSpringResourceContext ctx) throws Exception {
+        final Ignite ignite;
+
+        // if started node isn't first node in the local JVM then it was checked earlier for join to topology
+        // in IgniteProcessProxy constructor.
+        if (locJvmInstance == null && rmJvmInstance != null) {
+            final UUID syncId = ((IgniteProcessProxy)rmJvmInstance).getId();
+            final UUID nodeId = cfg.getNodeId();
+
+            ignite = super.startGrid(igniteInstanceName, cfg, ctx);
+
+            assert ignite.configuration().getNodeId() == nodeId : "Started node has unexpected node id.";
+
+            assert GridTestUtils.waitForCondition(new GridAbsPredicate() {
+                @Override public boolean apply() {
+                    return ignite.cluster().node(syncId) != null;
```

Comment:
may be add awaitPartitionMapExchange()

Reply:
There is a reverse synchronization in case of starting a the first node in local JVM when there are nodes in another JVM.

Reply:
By the same token, exchange doesn't start before cluster activation.